services/spar/src/Spar/Data.hs

```diff
@@ -0,0 +1,249 @@
+
+mkTTLAuthnRequests :: MonadError TTLError m => Env -> UTCTime -> m (TTL "authreq")
+mkTTLAuthnRequests (Env now maxttl _) = mkTTL now maxttl
+
+mkTTLAssertions :: MonadError TTLError m => Env -> UTCTime -> m (TTL "authresp")
+mkTTLAssertions (Env now _ maxttl) = mkTTL now maxttl
+
+mkTTL :: MonadError TTLError m => UTCTime -> TTL a -> UTCTime -> m (TTL a)
+mkTTL now maxttl endOfLife = if
+  | actualttl > maxttl -> throwError TTLTooLong
+  | actualttl <= 0     -> throwError TTLInPast
+  | otherwise          -> pure actualttl
+  where
+    actualttl = TTL . round @Double . realToFrac $ endOfLife `diffUTCTime` now
+
+err2err :: (m ~ Either TTLError, MonadThrow m') => m a -> m' a
+err2err = either (throwM . ErrorCall . show) pure
+
+
+----------------------------------------------------------------------
+-- saml state handling
+
+storeRequest :: (HasCallStack, MonadClient m) => Env -> SAML.ID SAML.AuthnRequest -> SAML.Time -> m ()
+storeRequest env (SAML.ID rid) (SAML.Time endOfLife) = do
+    TTL actualEndOfLife <- err2err $ mkTTLAuthnRequests env endOfLife
+    retry x5 . write ins $ params Quorum (rid, endOfLife, actualEndOfLife)
+  where
+    ins :: PrepQuery W (ST, UTCTime, Int32) ()
+    ins = "INSERT INTO authreq (req, end_of_life) VALUES (?, ?) USING TTL ?"
+
+checkAgainstRequest :: (HasCallStack, MonadClient m) => Env -> SAML.ID SAML.AuthnRequest -> m Bool
+checkAgainstRequest env (SAML.ID rid) = do
+    (retry x1 . query1 sel . params Quorum $ Identity rid) <&> \case
+        Just (Identity (Just endoflife)) -> endoflife >= dataEnvNow env
+        _ -> False
+  where
+    sel :: PrepQuery R (Identity ST) (Identity (Maybe UTCTime))
+    sel = "SELECT end_of_life FROM authreq WHERE req = ?"
+
+storeAssertion :: (HasCallStack, MonadClient m) => Env -> SAML.ID SAML.Assertion -> SAML.Time -> m Bool
+storeAssertion env (SAML.ID aid) (SAML.Time endOfLifeNew) = do
+    TTL actualEndOfLife <- err2err $ mkTTLAssertions env endOfLifeNew
+    notAReplay :: Bool <- (retry x1 . query1 sel . params Quorum $ Identity aid) <&> \case
+        Just (Identity (Just endoflifeOld)) -> endoflifeOld < dataEnvNow env
+        _ -> False
+    when notAReplay $ do
+        retry x5 . write ins $ params Quorum (aid, endOfLifeNew, actualEndOfLife)
+    pure notAReplay
+  where
+    sel :: PrepQuery R (Identity ST) (Identity (Maybe UTCTime))
+    sel = "SELECT end_of_life FROM authresp WHERE resp = ?"
+
+    ins :: PrepQuery W (ST, UTCTime, Int32) ()
+    ins = "INSERT INTO authresp (resp, end_of_life) VALUES (?, ?) USING TTL ?"
+
+
+----------------------------------------------------------------------
+-- user
+
+-- | Add new user.  If user with this 'SAML.UserId' exists, overwrite it.
+insertUser :: (HasCallStack, MonadClient m) => SAML.UserRef -> Brig.UserId -> m ()
+insertUser (SAML.UserRef tenant subject) uid = retry x5 . write ins $ params Quorum (tenant', subject', uid')
+  where
+    tenant', subject', uid' :: ST
+    tenant'  = cs $ SAML.encodeElem tenant
```

Comment:
How about having different types for `tenant`, `subject` and have `CQL` instances for it?

Reply:
Which do you mean?

1. change the type `SAML.UserRef` to have newtypes as components?  (that would mean the `CQL` instances would be orphans.)
2. change the implementation of `insertUser` to unwrap `UserRef` and then re-wrap its components?  I don't see the benefit of that, but I'm not strongly opposed either.
3. other?  (probably :-))

Reply:
Let's go for `1.` and add them to the existing ones under `Instances.hs`. Brig's uuid already has a CQL instance so there's no need to convert that one to text.